cmd/snap-confine/ns-support.c

```diff
@@ -549,137 +565,180 @@ int sc_create_or_join_mount_ns(struct sc_mount_ns *group,
-		// the namespace is pristine and needs to be populated now.
-		debug("unsharing the mount namespace");
-		if (unshare(CLONE_NEWNS) < 0) {
-			die("cannot unshare the mount namespace");
+		if (write(group->pipe_helper[1], &command, sizeof command) < 0) {
+			die("cannot write ack");
 		}
-		group->should_populate = true;
 	}
-	return 0;
+	debug("helper process exiting");
+	exit(0);
 }
 
-bool sc_should_populate_mount_ns(struct sc_mount_ns * group)
+static void helper_capture_ns(struct sc_mount_ns *group, pid_t parent)
 {
-	return group->should_populate;
+	char src[PATH_MAX] = { 0 };
+	char dst[PATH_MAX] = { 0 };
+
+	debug("capturing per-snap mount namespace");
+	sc_must_snprintf(src, sizeof src, "/proc/%d/ns/mnt", (int)parent);
+	sc_must_snprintf(dst, sizeof dst, "%s%s", group->name, SC_NS_MNT_FILE);
+	if (mount(src, dst, NULL, MS_BIND, NULL) < 0) {
+		die("cannot bind-mount the mount namespace file %s -> %s", src,
+		    dst);
+	}
+	debug("successfully captured mount namespace "
+	      "in namespace group %s", group->name);
 }
 
-void sc_preserve_populated_mount_ns(struct sc_mount_ns *group)
+static void sc_message_capture_helper(struct sc_mount_ns *group, int command_id)
 {
+	int ack;
 	if (group->child == 0) {
-		die("precondition failed: we don't have a support process for mount namespace capture");
+		die("precondition failed: we don't have a helper process");
 	}
-	if (group->event_fd < 0) {
-		die("precondition failed: we don't have an eventfd for mount namespace capture");
+	if (group->pipe_master[1] < 0) {
+		die("precondition failed: we don't have an pipe");
 	}
-	debug
-	    ("asking support process for mount namespace capture (pid: %d) to perform the capture",
-	     group->child);
-	if (eventfd_write(group->event_fd, 1) < 0) {
-		die("cannot write eventfd");
+	if (group->pipe_helper[0] < 0) {
+		die("precondition failed: we don't have an pipe");
 	}
-	debug
-	    ("waiting for the support process for mount namespace capture to exit");
+	debug("sending command %d to helper process (pid: %d)",
+	      command_id, group->child);
+	if (write(group->pipe_master[1], &command_id, sizeof command_id) < 0) {
+		die("cannot send command %d to helper process", command_id);
+	}
+	debug("waiting for response from helper");
+	if (read(group->pipe_helper[0], &ack, sizeof ack) < 0) {
+		die("cannot receive ack from helper process");
+	}
+}
+
+static void sc_wait_for_capture_helper(struct sc_mount_ns *group)
+{
+	if (group->child == 0) {
+		die("precondition failed: we don't have a helper process");
+	}
+	debug("waiting for the helper process to exit");
 	int status = 0;
 	errno = 0;
 	if (waitpid(group->child, &status, 0) < 0) {
-		die("cannot wait for the support process for mount namespace capture");
+		die("cannot wait for the helper process");
 	}
 	if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
-		die("support process for mount namespace capture exited abnormally");
+		die("helper process exited abnormally");
```

Comment:
Would it make sense to include exit status in the message here?

Reply:
The raw exit status can encode a number of cases (exit code, kill by signal) and I didn't want to make the distinction here originally. I can include it in a follow-up.